adbms/src/main/scala/de/up/hpi/informationsystems/adbms/definition/Relation.scala

```diff
@@ -55,9 +55,10 @@ trait Relation {
 
   /**
     * Inserts all Records into the relation.
+    * @note that this is not atomic
     * @param records to be inserted
     */
-  def insertAll(records: Seq[Record]): Unit = records.foreach(insert)
+  def insertAll(records: Seq[Record]): Try[Seq[Record]] = Success(records.map(insert).map(_.get))
```

Comment:
I guess, you want to do a `Try(records.map(insert).map(_.get))` to catch the case when one (or all) insert fails. Otherwise this method would throw the Throwable contained in `Failure[Throwable]` and not wrap it in the Failure.

You could even do `Try(records.map( r => insert(r).get ))` to stop at the first Failure (which will presumably be the first one in most of the cases). **Beware: could be over-engineering**

There is still no automatic rollback...